feat: Remote editor

dcalhoun opened the pull request:
Add experimental editor relying upon editor assets fetch from the remote WordPress site, enabling the ability to utilize third-party blocks.

To test: See https://github.com/wordpress-mobile/GutenbergKit/pull/22. 

## Regression Notes
1. Potential unintended areas of impact
  None, changed confined to feature hidden behind a disabled feature flag. 
2. What I did to test those areas of impact (or what existing automated tests I relied on)
  Verified previous editor worked as expected. 
4. What automated tests I added (or what prevented me from doing so)
  Deemed unnecessary. 

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

dcalhoun committed:
feat: Provide site URL to editor

Simplify top level site access.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: <code>Modules/Package.swift</c...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages in Xcode.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dcalhoun committed:
feat: Only use WPCOM API for Simple sites

Potentially simplify request authentication by relying upon the WP CORE
API for all requests -- both core endpoints and plugin-registered
endpoints.

dcalhoun committed:
build: Update GutenbergKit reference

dcalhoun committed:
feat: Add feature flag toggling block editor plugin support

dcalhoun committed:
feat: Manually retrieve blog password

It appears the application password is not currently stored when adding
a self-hosted site in the current login flow. This workaround change
should not be merged into the `trunk` branch.

dcalhoun committed:
build: Update GutenbergKit ref

dcalhoun committed:
build: Update GutenbergKit ref

dcalhoun commented on WordPress/Classes/ViewRelated/NewGutenberg/NewGutenbergViewController.swift:
```diff
@@ -134,12 +134,13 @@ class NewGutenbergViewController: UIViewController, PostEditor, PublishingEditor
 
         let selfHostedApiUrl = post.blog.url(withPath: "wp-json/")
-        let siteApiRoot = post.blog.isAccessibleThroughWPCom() ? post.blog.wordPressComRestApi()?.baseURL.absoluteString : selfHostedApiUrl
+        let siteApiRoot = post.blog.isAccessibleThroughWPCom() && post.blog.isHostedAtWPcom ? post.blog.wordPressComRestApi()?.baseURL.absoluteString : selfHostedApiUrl
```
Modified to force self-hosted sites to rely upon the core WP API, as the custom API endpoint returning editor assets resides on the core WP API path.

dcalhoun commented on WordPress/Classes/ViewRelated/NewGutenberg/NewGutenbergViewController.swift:
```diff
@@ -134,12 +134,13 @@ class NewGutenbergViewController: UIViewController, PostEditor, PublishingEditor
         var authHeader = "Bearer \(authToken)"
 
-        let applicationPassword = try? post.blog.getApplicationToken()
+        // TODO: Reinstate the application password token getter once supported
+        let applicationPassword = try? post.blog.getPassword()
```
From my testing, the application token is not utilized or available currently. I'm unsure as to the best approach for accessing the application token. Relates to https://github.com/wordpress-mobile/WordPress-iOS/pull/23499#discussion_r1749913408.

geriux replied:
Hey David 👋 

It is available under an experimental flag, here's how you can test it out:

**Precondition**: You have to be already logged in with a different account, because this feature is only available from the "My Sites" picker.

- Enable the feature flag by going to: Me > App Settings > Debug > Feature Flags > Enable: `Application Password for self-hosted sites`
- Go to the "My Site tab"
- Open the "My Sites" Picker
- Tap on the (+) button
- Tap on Add self-hosted site
- Enter your site address
- Follow the process of login into your site
- Approve the creation of the application password 

Once you finish the process the site should be added, and the application password ready to use.

> Relates to #23499 (comment).

I'm going to merge that PR now, but what I asked there was more related to code structure, if that was a good approach, or if there are better ways. Let's go with what we have and iterate if needed.

Let me know if you encounter any issues!

dcalhoun replied:
Thanks for sharing this context, @geriux. 

The fact that application password support is limited to adding sites from the site list is what caused me confusion. I had tried to use application passwords when logging in; i.e., adding my first site to the ap. Application passwords do appear to work correctly when adding a site to an existing list of sites. 

---

Tangentially related, I noted what appears to be an edge case when testing this now. If a targeted site's Jetpack account connection differs from the account used in the Jetpack mobile app, requests for the targeted site's content fail. Even though the site is added via application passwords, it appears requests are sent to the WPCOM API due to the presence of a Jetpack connection. However, the Jetpack connection in the mobile app is not associated with the targeted site and fails to authenticate. 

1. Create a site, connect Jetpack to **WPCOM Account A**. 
2. Log into the Jetpack mobile app with **WPCOM Account B**. 
3. Add the targeted self-hosted site via application passwords. 
4. 💥 All site requests are sent to `public-api.wordpress.com` and fail to locate the site.

dcalhoun commented on WordPress/Classes/ViewRelated/NewGutenberg/NewGutenbergViewController.swift:
```diff
@@ -149,7 +150,7 @@ class NewGutenbergViewController: UIViewController, PostEditor, PublishingEditor
         }
 
-        let siteApiNamespace = post.blog.dotComID != nil && applicationPassword == nil ? "sites/\(siteId ?? "")" : ""
+        let siteApiNamespace = post.blog.dotComID != nil && post.blog.isHostedAtWPcom && applicationPassword == nil ? "sites/\(siteId ?? "")" : ""
```
Modified to force self-hosted sites to rely upon the core WP API, as the custom API endpoint returning editor assets resides on the core WP API path.

dcalhoun commented on WordPress/Classes/ViewRelated/NewGutenberg/NewGutenbergViewController.swift:
```diff
@@ -160,6 +161,8 @@ class NewGutenbergViewController: UIViewController, PostEditor, PublishingEditor
             service: GutenbergKit.EditorService(client: networkClient),
             themeStyles: FeatureFlag.newGutenbergThemeStyles.enabled,
+            plugins: FeatureFlag.newGutenbergPlugins.enabled,
+            siteURL: post.blog.url ?? "",
```
Add to simplify loading the plugin's `remote.html` editor file in https://github.com/wordpress-mobile/GutenbergKit/pull/22.

dcalhoun committed:
fix: Limit remote editor to self-hosted sites

Currently, sites relying upon a WPCOM account utilize the WPCOM API and
an OAuth token. The remote editor requires a third-party endpoint on the
core WP API and an application token. Eventually, we hope to widen
support by landing the endpoint in Gutenberg/core.

dcalhoun commented on WordPress/Classes/ViewRelated/NewGutenberg/NewGutenbergViewController.swift:
```diff
@@ -160,6 +162,8 @@ class NewGutenbergViewController: UIViewController, PostEditor, PublishingEditor
             content: post.content ?? "",
             service: GutenbergKit.EditorService(client: networkClient),
             themeStyles: FeatureFlag.newGutenbergThemeStyles.enabled,
+            plugins: FeatureFlag.newGutenbergPlugins.enabled && isSelfHosted,
```
The remote editor is limited to self-hosted sites currently as it sites authenticating with a WPCOM account utilize the WPCOM API and an OAuth token. The remote editor requires a third-party endpoint found on the core WP API and an application password token (p1725915009777259/1725894719.476849-slack-C04PWEZSYFL).

dcalhoun committed:
Revert "feat: Manually retrieve blog password"

This reverts commit 1a28f0facca2758dafba0455f8238c4a5db3ec45.

dcalhoun committed:
build: Update GutenbergKit ref

geriux reviewed (approved):
LGTM!

geriux commented:
Once this lands we can add this new functionality in Android https://github.com/wordpress-mobile/WordPress-Android/pull/21223

dcalhoun committed:
build: Update GutenbergKit ref
